Add pants support for pytest

jasonwbarnett opened the pull request:
### Description

This is an example of how pants could be supported by the pytest runner.

### Context

https://github.com/buildkite/test-engine-client/issues/315

### Changes

- New `TestRunnerVariant` field added to config
- Validated to only allow "pants" if set
- Passed through to runner config and used in Pytest logic
- Skips Python package check and test discovery for pants variant
- Forces `BUILDKITE_TEST_ENGINE_TEST_COMMAND` to be set if variant is pants
- Logs helpful warnings and errors for unsupported behavior or misconfiguration

This enables integration with Pants without breaking existing workflows.

### Testing

I haven't deeply tested this yet as I wanted to open this to start a conversation on implementation details.

jasonwbarnett commented:
It works!

```text
$ export BUILDKITE_TEST_ENGINE_TEST_RUNNER=pytest
$ export BUILDKITE_TEST_ENGINE_TEST_RUNNER_VARIANT=pants
$ export BUILDKITE_TEST_ENGINE_TEST_CMD="pants test --output=all src/python/altana_devops/azure_devops/checks_test.py -- --json={{resultPath}}"
$ ~/git/test-engine-client/test-engine-client
+++ Buildkite Test Engine Client: bktec dev


______ ______ _____
___  /____  /___  /____________
__  __ \_  //_/  __/  _ \  ___/
_  /_/ /  ,<  / /_ /  __/ /__
/_.___//_/|_| \__/ \___/\___/

Info: Python package 'buildkite-test-collector' is required and will not be verified by bktec. Please ensure it is added to the pants resolve used by pytest.
Info: Pants test runner variant does not support discovering test files. Please ensure the test command is set correctly via BUILDKITE_TEST_ENGINE_TEST_COMMAND.
Billing Error: Test Splitting is not enabled on your plan: please upgrade at https://buildkite.com/organizations/altana/billing.
⚠️ Falling back to non-intelligent splitting. Your build may take longer than usual.
+++ Buildkite Test Engine Client: Running tests
pants test src/python/altana_devops/azure_devops/checks_test.py -- --json=/var/folders/kw/qhf2k3js3nvgf7t8_1g46skm0000gq/T/bktec-pytest-1365635794/pytest-results.json

Skipping automatic ecr-login because ALTANA_PANTS_SKIP_AUTO_ECR_LOGIN is set to true
16:40:22.93 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/altana_devops/azure_devops/checks_test.py:tests) - succeeded.

✓ src/python/altana_devops/azure_devops/checks_test.py:tests succeeded in 1.03s (ran in local environment `macos_arm64`).
+++ ========== Buildkite Test Engine Report  ==========
✅ All tests passed.

+---------+-----------+---+
| Passed  | first run | 6 |
+         +-----------+---+
|         | on retry  | 0 |
+---------+-----------+---+
| Muted   | passed    | 0 |
+         +-----------+---+
|         | failed    | 0 |
+---------+-----------+---+
| Failed  |           | 0 |
+---------+-----------+---+
| Skipped |           | 0 |
+---------+-----------+---+
|               TOTAL | 6 |
+---------+-----------+---+
===================================================
```

jasonwbarnett commented:
It works!

```text
$ export BUILDKITE_TEST_ENGINE_TEST_RUNNER=pytest
$ export BUILDKITE_TEST_ENGINE_TEST_RUNNER_VARIANT=pants
$ export BUILDKITE_TEST_ENGINE_TEST_CMD="pants test --output=all src/python/devops/azure_devops/checks_test.py -- --json={{resultPath}}"
$ ~/git/test-engine-client/test-engine-client
+++ Buildkite Test Engine Client: bktec dev


______ ______ _____
___  /____  /___  /____________
__  __ \_  //_/  __/  _ \  ___/
_  /_/ /  ,<  / /_ /  __/ /__
/_.___//_/|_| \__/ \___/\___/

Info: Python package 'buildkite-test-collector' is required and will not be verified by bktec. Please ensure it is added to the pants resolve used by pytest.
Info: Pants test runner variant does not support discovering test files. Please ensure the test command is set correctly via BUILDKITE_TEST_ENGINE_TEST_COMMAND.
Billing Error: Test Splitting is not enabled on your plan: please upgrade at https://buildkite.com/organizations/<redacted>/billing.
⚠️ Falling back to non-intelligent splitting. Your build may take longer than usual.
+++ Buildkite Test Engine Client: Running tests
pants test src/python/devops/azure_devops/checks_test.py -- --json=/var/folders/kw/qhf2k3js3nvgf7t8_1g46skm0000gq/T/bktec-pytest-1365635794/pytest-results.json

16:40:22.93 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/checks_test.py:tests) - succeeded.

✓ src/python/devops/azure_devops/checks_test.py:tests succeeded in 1.03s (ran in local environment `macos_arm64`).
+++ ========== Buildkite Test Engine Report  ==========
✅ All tests passed.

+---------+-----------+---+
| Passed  | first run | 6 |
+         +-----------+---+
|         | on retry  | 0 |
+---------+-----------+---+
| Muted   | passed    | 0 |
+         +-----------+---+
|         | failed    | 0 |
+---------+-----------+---+
| Failed  |           | 0 |
+---------+-----------+---+
| Skipped |           | 0 |
+---------+-----------+---+
|               TOTAL | 6 |
+---------+-----------+---+
===================================================
```

jasonwbarnett commented:
Since opening this PR, I’ve realized there’s a key issue that needs to be accounted for:

1. test-engine-client assumes a single invocation of pytest, and therefore only a single {{resultPath}}.
This doesn’t work with Pants (unless using xdist), because Pants spawns a separate pytest process per test—or per group of tests when using [batch_compatibility_tag][batch-compatibility-tag]. Each of these processes writes to the same JSON file, which results in only one set of test results being preserved. The others are overwritten.

One potential solution would be to update [test-collector-python] to use a [filelock][filelock] and append or merge results into the JSON file in a thread/process-safe way. This would allow multiple processes to contribute to a single result file without clobbering each other, and still keep things compatible with test-engine-client.

nprizal commented:
Hi @jasonwbarnett. Thank you for opening the PR. I might be missing some contexts here, but can you explain more about the support you want to add, or the use case you might have? 

`bktec` is currently designed around test splitting, where multiple instances of the _client_ run a subset of test files that it receives from the server as a _Test Plan_. However, looking at your changes, it seems like _bktec_ won't do the splitting. Instead, `pants` will spawn multiple pytest processes, which is not what `bktec` is designed currently.

We do have other features, such as muting or retrying tests. If that's something you'd want to achieve, we could add the support, but but it might require a slightly different approach compared to other test runners with splitting support.

jasonwbarnett commented:
Exactly! I'm looking for the muting and retry features. Test splitting simply isn't possible given our use case and pants which has much more information. Not to say it's impossible it's just not a priority for us at this time. We are desperate to get the ability to automatically ignore flaky tests and have developers be able to mark tests as flaky manually so we can ignore the results of those tests.

jasonwbarnett commented:
@nprizal Exactly! We’re primarily interested in the muting and retry features. Test splitting isn’t feasible in our case due to Pants, which already has richer context about the test structure. It’s not that it couldn’t be done, it’s just not a priority for us right now. What we’re really eager for is the ability to automatically mute flaky tests and allow developers to manually mark tests as flaky so their results can be excluded.

nprizal commented:
@jasonwbarnett Got it! After looking at your changes, it seems like the runner behaves slightly differently when using `pants`. The main thing that is the same is the results handling. Personally, I prefer to create a separate runner and reuse some of the shared functionality. This will make it easier if we want to implement a splitting feature for pants.

```go
// pytest.go
type Pytest struct {}

func NewPytest(c Config) Pytest {
  ...
}

func (p Pytest) GetFiles() (string[], error) {
  ...
}

func (p Pytest) Run(...) error {
  ...
  err = runAndForwardSignal(cmd)
  ...
  err = HandlePytestResult(result, p.ResultPath)
  ...
}

func HandlePytestResult(result RunResult, path string) error { // this is the shared function
   tests, err := ParsePytestCollectorResult(p.ResultPath)
   ...
   for _, test := range tests {
     result.RecordTestResult(...)
   }
   ...
} 
```

```go
// pytest_pants.go
type PytestPants struct {}

func NewPytestPants(c Config) {
  ...
}

func (p PytestPants) GetFiles() (string[], error) {
  return []string{}, nil
}

func (p PytestPants) Run(...) error {
  ...
  err = runAndForwardSignal(cmd)
  ...
  err = HandlePytestResult(result, p.ResultPath)
  ...
}
```
We are also planning of extracting the results handling, including the parsing, separately from the runner in the future. So that the `runner` package is responsible for running the tests, while the `result_handler` is responsible for parsing and recording the results.

```
internal
  /runner
    /pytest.go
    /pytest_pants.go
  /result_handler
    /pytest.go
    /run_result.go
    /test_result.go
```

> One potential solution would be to update [test-collector-python] to use a [filelock][filelock] and append or merge results into the JSON file in a thread/process-safe way. This would allow multiple processes to contribute to a single result file without clobbering each other, and still keep things compatible with test-engine-client.

This sounds like a sensible approach. My main concern is that in the current `bktec` pytest runner implementation, we create a temporary directory to store the test result if the `BUILDKITE_TEST_ENGINE_RESULT_PATH` is not set. This directory will be destroyed when `bktec` terminated. We can make the `BUILDKITE_TEST_ENGINE_RESULT_PATH` mandatory for `pants` so we can use that file for many pytest processes.

jasonwbarnett commented:
@nprizal I've updated the PR and implemented pytest pants runner distinct and separate from the pytest runner. It shares the parsing logic (just reused it). This will not work unless https://github.com/buildkite/test-collector-python/pull/55 is merged.

jasonwbarnett commented:
Tests are passing locally:

```text
Running tool: /opt/homebrew/bin/go test -timeout 30s -run ^(TestPytestPantsRun|TestPytestPantsRun_RetryCommand|TestPytestPantsRun_TestFailed|TestPytestPantsRun_CommandFailed|TestPytestPantsGetFiles|TestPytestPantsGetExamples|TestPytestPantsCommandNameAndArgs_WithResultPath|TestPytestPantsCommandNameAndArgs_WithoutResultPath|TestPytestPantsCommandNameAndArgs_InvalidTestCommand)$ github.com/buildkite/test-engine-client/internal/runner

ok  	github.com/buildkite/test-engine-client/internal/runner	2.420s
```

jasonwbarnett commented:
I just finished adding documentation to this PR. I'm a little confused what to label and name things because pants can technically run many different types of tests, i.e. python, javascript, kotlin, golang, scala, typescript, etc

This PR is strictly aiming to support Python which is arguably the most predominantly used language in pants.

I'm open to changing the naming of things as you see fit.

Cheers!

jasonwbarnett commented:
I confirmed the --merge-json flag works as expected.

---

## with --merge-json

```text
$ export BUILDKITE_TEST_ENGINE_TEST_CMD="pants --filter-target-type=python_test test //src/python/devops/azure_devops:: -- --json={{resultPath}} --merge-json"
$ ~/git/test-engine-client/test-engine-client
+++ Buildkite Test Engine Client: bktec dev


______ ______ _____
___  /____  /___  /____________
__  __ \_  //_/  __/  _ \  ___/
_  /_/ /  ,<  / /_ /  __/ /__
/_.___//_/|_| \__/ \___/\___/

Info: Python package 'buildkite-test-collector' is required and will not be verified by bktec. Please ensure it is added to the pants resolve used by pytest.
2025/06/06 14:08:02 DEBUG: Fetching test plan
2025/06/06 14:08:02 DEBUG: Sending request GET https://api.buildkite.com/v2/analytics/organizations/org/suites/test-out-new-thing/test_plan?identifier=<redacted>/<redacted>&job_retry_count=0
2025/06/06 14:08:03 DEBUG: Response code 404
2025/06/06 14:08:03 DEBUG: No test plan found, creating a new plan
2025/06/06 14:08:03 DEBUG: Test plan created. Identifier: "<redacted>/<redacted>"
2025/06/06 14:08:03 DEBUG: My favourite ice cream is
+++ Buildkite Test Engine Client: Running tests
pants --filter-target-type=python_test test //src/python/devops/azure_devops:: -- --json=/var/folders/kw/qhf2k3js3nvgf7t8_1g46skm0000gq/T/bktec-pytest-189074465/pytest-results.json --merge-json

Skipping automatic ecr-login because PANTS_SKIP_AUTO_ECR_LOGIN is set to true
14:08:16.89 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/checks_test.py:tests) - succeeded.
14:08:16.90 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/pull_request_description_test.py:tests) - succeeded.
14:08:16.99 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/releases/notes/helpers_test.py:tests) - succeeded.

✓ src/python/devops/azure_devops/checks_test.py:tests succeeded in 1.47s (ran in local environment `macos_arm64`).
✓ src/python/devops/azure_devops/pull_request_description_test.py:tests succeeded in 1.47s (ran in local environment `macos_arm64`).
✓ src/python/devops/azure_devops/releases/notes/helpers_test.py:tests succeeded in 1.56s (ran in local environment `macos_arm64`).
+++ ========== Buildkite Test Engine Report  ==========
✅ All tests passed.

+---------+-----------+----+
| Passed  | first run | 18 |
+         +-----------+----+
|         | on retry  |  0 |
+---------+-----------+----+
| Muted   | passed    |  0 |
+         +-----------+----+
|         | failed    |  0 |
+---------+-----------+----+
| Failed  |           |  0 |
+---------+-----------+----+
| Skipped |           |  0 |
+---------+-----------+----+
|               TOTAL | 18 |
+---------+-----------+----+
===================================================
```
---

### With*out* --merge-json

```text
$ export BUILDKITE_TEST_ENGINE_TEST_CMD="pants --filter-target-type=python_test test //src/python/devops/azure_devops:: -- --json={{resultPath}}"
$ ~/git/test-engine-client/test-engine-client
+++ Buildkite Test Engine Client: bktec dev


______ ______ _____
___  /____  /___  /____________
__  __ \_  //_/  __/  _ \  ___/
_  /_/ /  ,<  / /_ /  __/ /__
/_.___//_/|_| \__/ \___/\___/

Info: Python package 'buildkite-test-collector' is required and will not be verified by bktec. Please ensure it is added to the pants resolve used by pytest.
2025/06/06 14:08:02 DEBUG: Fetching test plan
2025/06/06 14:08:02 DEBUG: Sending request GET https://api.buildkite.com/v2/analytics/organizations/org/suites/test-out-new-thing/test_plan?identifier=<redacted>/<redacted>&job_retry_count=0
2025/06/06 14:08:03 DEBUG: Response code 404
2025/06/06 14:08:03 DEBUG: No test plan found, creating a new plan
2025/06/06 14:08:03 DEBUG: Test plan created. Identifier: "<redacted>/<redacted>"
2025/06/06 14:08:03 DEBUG: My favourite ice cream is
+++ Buildkite Test Engine Client: Running tests
pants --filter-target-type=python_test test //src/python/devops/azure_devops:: -- --json=/var/folders/kw/qhf2k3js3nvgf7t8_1g46skm0000gq/T/bktec-pytest-189074465/pytest-results.json --merge-json

Skipping automatic ecr-login because PANTS_SKIP_AUTO_ECR_LOGIN is set to true
14:08:16.89 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/checks_test.py:tests) - succeeded.
14:08:16.90 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/pull_request_description_test.py:tests) - succeeded.
14:08:16.99 [INFO] Completed: Run Pytest - (environment:macos_arm64, src/python/devops/azure_devops/releases/notes/helpers_test.py:tests) - succeeded.

✓ src/python/devops/azure_devops/checks_test.py:tests succeeded in 1.47s (ran in local environment `macos_arm64`).
✓ src/python/devops/azure_devops/pull_request_description_test.py:tests succeeded in 1.47s (ran in local environment `macos_arm64`).
✓ src/python/devops/azure_devops/releases/notes/helpers_test.py:tests succeeded in 1.56s (ran in local environment `macos_arm64`).
+++ ========== Buildkite Test Engine Report  ==========
✅ All tests passed.

+---------+-----------+----+
| Passed  | first run | 6 |
+         +-----------+----+
|         | on retry  |  0 |
+---------+-----------+----+
| Muted   | passed    |  0 |
+         +-----------+----+
|         | failed    |  0 |
+---------+-----------+----+
| Failed  |           |  0 |
+---------+-----------+----+
| Skipped |           |  0 |
+---------+-----------+----+
|               TOTAL | 6 |
+---------+-----------+----+
===================================================
```

---

nprizal commented on internal/runner/detector.go:
```diff
@@ -47,10 +47,12 @@ func DetectRunner(cfg config.Config) (TestRunner, error) {
 		return NewPlaywright(runnerConfig), nil
 	case "pytest":
 		return NewPytest(runnerConfig), nil
+	case "pytestpants":
```
I don't have strong opinion on the naming, but maybe we can split this 2 words `pytest-pants`.

jasonwbarnett replied:
Finished.

nprizal reviewed (changes requested):
Overall looks good. I have comments around the naming and the test plan creation. I have also turned on the 3rd party contributor build, so next time you push a commit, it should trigger a CI build in the pipeline.

jasonwbarnett commented on main.go:
```diff
@@ -361,19 +361,25 @@ func fetchOrCreateTestPlan(ctx context.Context, apiClient *api.Client, cfg confi
-	debug.Println("Creating test plan")
-	testPlan, err := apiClient.CreateTestPlan(ctx, cfg.SuiteSlug, params)
+	var testPlan plan.TestPlan
+	if params.Runner != "pytestpants" {
```
I found where it's building params for the request and just swapped it with `pytest`. Let me know if I should do something else here.

jasonwbarnett commented:
@nprizal Thanks for the support. It looks like the tests are passing! 🎉 🎉

jasonwbarnett commented on internal/runner/pytest_pants.go:
```diff
@@ -0,0 +1,141 @@
+
+func (p PytestPants) commandNameAndArgs(cmd string, testCases []string) (string, []string, error) {
+	if strings.Contains(cmd, "{{testExamples}}") {
+		return "", []string{}, fmt.Errorf("currently, bktec does not support dynamically injecting {{testExamples}}. Please ensure the test command in BUILDKITE_TEST_ENGINE_TEST_COMMAND does *not* include {{testExamples}}")
```
Woopsie! It should be BUILDKITE_TEST_ENGINE_TEST_CMD. Fixing now.

jasonwbarnett replied:
Pushed a fix.

nprizal commented on main.go:
```diff
@@ -393,15 +393,21 @@ func createRequestParam(ctx context.Context, cfg config.Config, files []string,
-		}, nil
+		}
+
+		if cfg.TestRunner == "pytest-pants" {
```
We can maybe add a comment to explain why this is necessary?

nprizal commented on docs/pytest-pants.md:
```diff
@@ -0,0 +1,82 @@
+In both commands, `{{resultPath}}` is replaced with a unique temporary path created by bktec. `--json` option is a custom pytest option added by Buildkite Test Collector to save the result into a JSON file at given path. You can further customize the test command for your specific use case.
+
+> [!IMPORTANT]
+> Make sure to append `-- --json={{resultPath}}` in your custom pants test command, as bktec requires this to read the test results for retries and verification purposes.
```
I believe the `--merge-json` is also required for pants
```suggestion
> Make sure to append `-- --json={{resultPath}} --merge-json` in your custom pants test command, as bktec requires this to read the test results for retries and verification purposes.
```

nprizal reviewed (approved):
This looks great 🚀! The changes are isolated, and won't affect the existing `pytest` runner and other common behavior. I have few non blocking comments.

jasonwbarnett commented:
> This looks great 🚀! The changes are isolated, and won't affect the existing `pytest` runner and other common behavior. I have few non blocking comments.

Thanks! I'll get to this in an hour.

jasonwbarnett commented:
@nprizal I've updated the code with the two changes. Should I squash all commits or will you do that on merge?

nprizal reviewed (approved):
@jasonwbarnett Thanks for adding the comment! Could you please squash the commits? Once you’ve done that, I’ll merge the PR and create a _release candidate_ so your team can begin testing it.

jasonwbarnett commented:
@nprizal squashed and pushed.

nprizal commented:
@jasonwbarnett I have created a release candidate that includes this PR
https://github.com/buildkite/test-engine-client/releases/tag/v1.6.0-rc.1

nprizal commented:
@jasonwbarnett I think you can still edit the PR description.
